### ISSUE TITLE
Release NCHS Mortality

### ISSUE THREAD

**Issue by krivard** [cmu-delphi/covidcast-indicators#367]
- [Link to issue](https://github.com/cmu-delphi/covidcast-indicators/issues/119)
- [Link to PR](https://github.com/cmu-delphi/covidcast-indicators/pull/233)
- Proposed release version: 1.11

This is a new indicator that's completed its first phase of development and is ready to consider for public release. This indicator tracks the number of covid, pneumonia, and all deaths, as well as the percentage of expected deaths, as published by the NCHS. Signal is per week, but the estimate is updated daily.

**Edit: we will not pursue map release at this time**

- [x] Statistical review (usually [correlations](https://github.com/cmu-delphi/covidcast/tree/main/docs/R-notebooks))
- [x] Signal / source name review (usually [Roni](https://docs.google.com/document/d/10hGd4Evce4lJ4VkWaQEKFQxvmw2P4xyYGtIAWF52Sf8/edit?usp=sharing))
- [x] Add to Delphi Automation
- [x] API support for weekly data

- [x] Visual review
- [ ] ~[Signal description pop-up text](https://docs.google.com/document/d/1kDqRg8EaI4WQXMaUUbbCGPlsUqEql8kgXCNt6AvMA9I/edit?usp=sharing) drafting and review~
- [ ] ~[Map release notes](https://docs.google.com/document/d/1BpxGgIma_Lkd2kxtwEo2DBdHQ3zk6dHRz-leUIRlOIA/edit?usp=sharing)~

- [x] API [documentation](https://github.com/cmu-delphi/delphi-epidata/tree/main/docs/api) and/or [changelog](https://github.com/cmu-delphi/delphi-epidata/blob/main/docs/api/covidcast_changelog.md)
- [ ] API mailing list notification

**Comment by benjaminysmith** [comment]
Thanks Katie -- would you like me to shepherd the sub-tasks? Do you have an example of the process I can copy?

**Comment by krivard** [comment]
I would! Some examples are in the links, but this is still a prototype of the process we'd like to build, so not everything has something to go off of.

The tasks are roughly in order -- eg if the statistical review doesn't go off, we don't want to move forward until that's been fixed.

Jingjing should handle the statistical review. Anyone can handle putting together the signal and source naming recommendations (just send Roni a txt when he should look at the approvals doc and when it's needed by -- 1.11 isn't on the calendar yet so several days allowance is fine). Adding to automation will coordinate with Brian. Visual review, signal description pop-up text, and map release notes will coordinate with Chris. API documentation and mailing list notification will coordinate with Kari, Jingjing, and Alex.

**Comment by benjaminysmith** [comment]
Fabulous thanks!

@jingjtang looks like you are up for the first task. Ping me if you need help and I can escalate.

**Comment by jingjtang** [comment]
Simple comparison between `usa-facts` `deaths_incidence_num` vs `nchs` `wip_covid_deaths_num`. They are consistent each other in general but difference exists. The extent of difference varies among states.
[covid_deaths_num.pdf](https://github.com/cmu-delphi/covidcast-indicators/files/5446811/covid_deaths_num.pdf)

As for the correlation APP,  it seems there is not a weekly response to use. A simple geo-wise spearman correlation analysis is shown below. Note that there is no lag considered(according to the figures in the pdf). The drop in the last few weeks is reasonable, since deaths usually have significant delay in reports. According to [NCHS](https://www.cdc.gov/nchs/nvss/vsrr/COVID19/index.htm), this delay can range from 1 week to 8 weeks or more, depending on the jurisdiction and cause of death.
![image](https://user-images.githubusercontent.com/31444565/97340986-db483d80-185a-11eb-895c-0f9ef6bc5671.png)

**Comment by jingjtang** [comment]
@krivard As for the signal names, the current ones are:
- `covid_deaths_num`
- `covid_deaths_prop`
- `total_deaths_num`
- `total_deaths_prop`
- `influenza_deaths_num`
- `influenza_deaths_prop`
- `pneumonia_deaths_num`
- `pneumonia_deaths_prop`
- `pneumonia_and_covid_deaths_num`
- `pneumonia_and_covid_deaths_prop`
- `pneumonia_influenza_or_covid_19_deaths_num`
- `pneumonia_influenza_or_covid_19_deaths_prop`
- `percent_of_expected_deaths`


as described [here](https://github.com/cmu-delphi/covidcast-indicators/blob/nchs_mortality/nchs_mortality/DETAILS.md#metrics-level-1-m1)

**Comment by jingjtang** [comment]
Though there is no geo aggregation needed, but we need population info for `prop` signals, so still need the geo refactor. However, state names (e.g. `Alabama`, `Alaska`) are used in the raw dataset which has not been supported by our geo utils. Need @dshemetov 's confirm.

**Comment by jingjtang** [comment]
Contacted @korlaxxalrok already for the automation. Will work with him once he starts on this. 
As for the automation, some important points for this pipeline
- [ ] A token is needed. Either use mine or create another one for Delphi.
- [ ] Code related to `export_start_date` should be changed according to the final strategy
- [ ] Data versioning staffs (run daily but store diffs weekly), might need @eujing 's help.

**Comment by RoniRos** [comment]
@jingjtang : 

1. Is there significance to the use of 'and' vs. 'or' in the signal names above?  Does `pneumonia_and_covid_deaths_num`  count deaths that have been classified as due to BOTH pneumonia AND Covid, or those that have been classified as due to Influenza **+** those classified as due to Covid?

2. In [this description](https://github.com/cmu-delphi/covidcast-indicators/blob/nchs_mortality/nchs_mortality/DETAILS.md#metrics-level-1-m1), it's important to clearly mention whether the classification is based only on the primary ICD code (primary cause of death), or on 'any code'.

**Comment by jingjtang** [comment]
@RoniRos 

1. If we consider deaths caused by `influenza`, `pneumonia` and `covid` as different sets. `and` means `Intersection` while `or` means `union`. So, 
   - `pneumonia_and_covid_deaths_num` means deaths caused by BOTH `pneumonia` AND `covid`.   
   - `pneumonia_influenza_or_covid_19_deaths` means deaths caused by EITHER `influenza`, `pneumonia` OR `covid`

2. As described in their [website](https://www.cdc.gov/nchs/nvss/vsrr/COVID19/index.htm). I think they are classified based on 'any code'. (not sure whether it is the primary cause) 
    `COVID-19 deaths are identified using a new ICD–10 code. When COVID-19 is reported as a cause of death – or when it is listed as a “probable” or “presumed” cause — the death is coded as U07.1. This can include cases with or without laboratory confirmation.`

**Comment by jingjtang** [comment]
@korlaxxalrok for automation

**Comment by benjaminysmith** [comment]
@RoniRos -- is the clarification on naming acceptable, or would you like to change the AND/OR use?

@krivard is there an approver for the correlations above, or is this enough?

**Comment by jingjtang** [comment]
API documentation PR [here ](https://github.com/cmu-delphi/delphi-epidata/pull/264#issue-511020952)

**Comment by benjaminysmith** [comment]
The correlation looks pretty convincing to me. @RoniRos can give final approval on this, but otherwise this looks good to go.

If the names are also good I think we have everything we need aside from running it in automation.

**Comment by jingjtang** [comment]
Visual View for all of the signals [here](https://drive.google.com/file/d/1tuIdRg4JxbXYVoDHDIhbUnvOtUKJL0zK/view?usp=sharing)

**Comment by RoniRos** [comment]
@benjaminysmith I'm very sorry this has been blocked on me.  I needed to find time to review the NCHS data definitions.
@jingjtang : Are the proposed names above supposed to match to all the columns [here](https://www.cdc.gov/nchs/nvss/vsrr/COVID19/index.htm), except excluding the "Percent of Expected Deaths"?

**Comment by RoniRos** [comment]
Based on the NCHS explanations and our conventions, the signals need to named as follows:

- deaths_covid_cumulative_num
- deaths_covid_cumulative_prop
- deaths_allcause_cumulative_num
- deaths_allcause_cumulative_prop
- deaths_flu_cumulative_num
- deaths_flu_cumulative_prop
- deaths_pneumonia_butnotflu_cumulative_num
- deaths_pneumonia_butnotflu_cumulative_prop
- deaths_covid_and_pneumonia_butnotflu_cumulative_num
- deaths_covid_and_pneumonia_butnotflu_cumulative_prop
- deaths_pneumonia_or_flu_or_covid_cumulative_num
- deaths_pneumonia_or_flu_or_covid_cumulative_num
- deaths_percent_of_expected

By way of explanation: it turns out the classifications are _not _ based on the primary cause of death, but on any cause of death, of which there are often several (several ICD codes per individual death).  So the logic has to be spelled out more clearly.

Some choices can be debated.  I'd love to hear your thoughts @krivard @benjaminysmith @jingjtang  : 

* I started with 'deaths' because that's what we do in the existing Covid death signals (although we don't use 'covid' in their name, it is implied).
* The example on the CDC website shows cumulative counts, so I made it 'cumulative'.  If the signal is not cumulative, then 'cumulative' should be replaced by 'incidence' (or do you provide both signals?).
* 'total' is too generic a term - there are many ways of deriving totals.  The technical term is 'all cause', which we could spell as 'all_cause' or 'allcause' or even 'all-cause' (@krivard do we ever use a hyphen in signal names?).
* 'flu' and 'influenza' are generally equivalent.  In formal writing, 'influenza' is usually preferred, because 'flu' in general public usage is used for lots of other things (e.g. "stomach flu", which has nothing to do with influenza).  However, these names are awfully long already, so I think it is acceptable and indeed preferable to use 'flu'.
* "butnotflu" could be spelled "but_not_flu", if you prefer.  Again, I was going for simplicity.

**Comment by RoniRos** [comment]
> In this description, it's important to clearly mention whether the classification is based only on the primary ICD code (primary cause of death), or on 'any code'.

@jingjtang please modify the description to explicitly state that the classification is based on all the codes on the death certificate (not just the 'primary cause of death').

**Comment by RoniRos** [comment]
> The correlation looks pretty convincing to me. @RoniRos can give final approval on this,
@benjaminysmith Can you please point me to the correlations?

**Comment by jingjtang** [comment]
@RoniRos  They should be `incidence` not `cumulative`. You can see a simple comparison between the NCHS `covid deaths_num` and usa-facts `deaths_incidence_num` [here](https://github.com/cmu-delphi/covidcast-indicators/issues/367#issuecomment-717415555)

**Comment by RoniRos** [comment]
> @RoniRos They should be `incidence` not `cumulative`. You can see a simple comparison between the NCHS `covid deaths_num` and usa-facts `deaths_incidence_num` [here](https://github.com/cmu-delphi/covidcast-indicators/issues/367#issuecomment-717415555). Currently we only provide `incidence`, we can add `cumulative` ones if we want.

Ok, thanks!   Then let's change to 'incidence' throughout.

Yes, I saw this plot, I thought Ben was referring to additional ones.
I'd love to view all these signals before we make them public.  How can I do that?

**Comment by jingjtang** [comment]
@RoniRos This is not automated yet and the signals cannot be fetched from our API. I am not sure what's the best way to visually show them. [Here](https://github.com/cmu-delphi/covidcast-indicators/issues/367#issuecomment-722723456) is a notebook that I created for you to play with.

**Comment by RoniRos** [comment]
Thank you!  That worked.

PS: if you are going to keep the notebook, you may want to revise the section titles, because they do not currently reflect the actual case definitions.  You can take the titles from the NCHS column headings, e.g. "Pneumonia and COVID Deaths" should actually be "Death involving COVID or pneumonia but not flu", etc.

**Comment by krivard** [comment]
@RoniRos I've copied your proposed names into [the approvals doc ](https://docs.google.com/document/d/10hGd4Evce4lJ4VkWaQEKFQxvmw2P4xyYGtIAWF52Sf8/edit?usp=sharing) for easier discussion. 

Hyphens are not permitted in signal names, only underscores.

Signal names can be up to 64 characters long; prefer limiting to 60 in case we need to use the old `wip_` prefix system (though we're very close to having a separate staging environment where we can just use the final system names without prefixing). I've annotated all the names with their length.

**Comment by jingjtang** [comment]
- Pipeline updated in the new PR #456 
- API documentation updated [here](https://github.com/cmu-delphi/delphi-epidata/pull/264)

**Comment by benjaminysmith** [comment]
Thanks Roni and Jingjing!

It looks like signal names are approved in the doc and updated in the documentation PR.

@RoniRos -- any final concerns from you on the correlations or visual review from the notebook, or can we mark those as approved?

**Comment by RoniRos** [comment]
No concerns about the correlations per se.  But I note that correlations are weak for 2020w11 and w12, are intermediate for w13, and are appropriately high starting only from w14.  @jingjtang @krivard do you know which of the two sources (NCHS and USA Facts) is "wrong"?   Depending on the answer, we may want to start the NCSH signals on w11 or w14.

**Comment by jingjtang** [comment]
@RoniRos According to the comparison in time series plots ([compare_with_usafacts.pdf](https://github.com/cmu-delphi/covidcast-indicators/files/5519388/compare_with_usafacts.pdf)), there is almost always a 1-2 (or even more) delay in NCHS compared with USA-Facts. The first one is based on death certificates while it is not very clear where the death counts come from. USA-Facts states that they complied their data from local government either through scraping or manual entry. I searched for some of the government websites (MA, PA, AL), only AL clearly explained under their [COVID dashboard](https://alpublichealth.maps.arcgis.com/apps/opsdashboard/index.html#/6d2771faa9da4a2786a509d82c8cf0f7) that the death counts are based on death certificates with COVID as the _**underlying cause of death**_. (And there isn't an obvious shift in AL from W11 to W14 compared with other states.)

Reconsider the 'primary cause problem':
[Here](https://www.cdc.gov/nchs/nvss/vsrr/covid19/excess_deaths.htm), CDC explained that 
```
Cause of death counts are based on the underlying cause of death, and presented for Respiratory diseases, Circulatory diseases, Malignant neoplasms, and Alzheimer disease and dementia. 
```
And the `Daily Updates of Totals by Week and State` that we used for our `nchs_mortality` pipeline is a content of this topic, which might show that my previous statement is **_wrong_**. But the explanation on the website for [Table 1](https://www.cdc.gov/nchs/nvss/vsrr/covid19/index.htm)(where we direct get the provisional death data), is shown like this
```
COVID-19 deaths are identified using a new ICD–10 code. When COVID-19 is reported as a cause of death – or when it is listed as a “probable” or “presumed” cause — the death is coded as U07.1. This can include cases with or without laboratory confirmation.
```
It confuses me. I think we need re-consider the definition.

**Comment by benjaminysmith** [comment]
@jingjtang -- what do we need to do to resolve this? Do you need additional input?

**Comment by RoniRos** [comment]
Thanks @jingjtang .  

1. I wasn't concerned about reporting delays, nor about the explanation being inconsistent.  I was referring to the  observation that in your geo-wise correlation plot:
![image](https://user-images.githubusercontent.com/23179630/99157584-2c985f80-2698-11eb-85ae-0a2df4480fa9.png) 
The correlation is appropriately high (almost 1.0) for most of the weeks, but not for the first few weeks (w11-w12 and maybe w13).  To me this suggests that at either the NCHS or the USAFacts signal have wrong data for these weeks, and I was asking which.  I don't think that reporting delays in NCHS make a difference here, because by now there are no more updates to NCHS reports for w11-14.  And NCHS data is based on the date of death, not the date of reporting.  (@jingjtang : Does USAFacts  report deaths by date of death or by the date the death was reported?)

On inspecting your comparison chart ([covid_deaths_num.pdf](https://github.com/cmu-delphi/covidcast-indicators/files/5446811/covid_deaths_num.pdf)).  I think I figured out the reason.  There are some surprising zeros in the NCHS counts.  Specifically:
* Alaska (AK) is all zeros, even though Table 2 [here ](https://www.cdc.gov/nchs/nvss/vsrr/COVID19/index.htm) shows a cumulative 74 COVID deaths.
* HI is zero thru w33
* MT is zero thru w31
* Quite a few states are zero thru w14 or w15, even though USAFacts reports cases 

So @benjaminysmith I think my question about the glitchy correlation is resolved: the problem is NCHS, not USAFacts.  But this is probably something we should try to resolve.  For example, why does our NCHS signal reports all zeros for AK when the CDC table shows cumulative 74 deaths?  And are the prolonged zeros for HI, MT & others correspond to CDC's reporting?

**Comment by krivard** [comment]
Is it likely that this is related to the file format problems with the early NCHS data files dated up to ~May? I inquired with Matthew but never got a response, and we eventually just went forward with the better file format that became available starting June. [(Sparse) context in this thread.](https://github.com/cmu-delphi/covidcast-indicators/issues/119#issuecomment-661263330)

**Comment by RoniRos** [comment]
The weak correlations for w11-13 may be explained that way, but not the zeros for AK, WY,HI,MT.  I think we need to resolve this before we publish -- we shouldn't publish zeros for AK, for example.
@jingjtang can you please investigate why e.g. AK is still showing zeros, even though the NCHS table shows cumulative 74 deaths?  Please point me to the location in NCHS where the zeros originate.
@krivard by Matthew do you mean Matt Biggerstaff at CDC?  I now have a more direct connection to NCHS, so I could try with them, but before I do I want to make sure I have all the fact right.

**Comment by jingjtang** [comment]
@RoniRos  Just fixed a bug in the pipeline that I incorrect put a one week shift previously #528. After fixing this issue:
- The current comparison is shown like this 
[compare_with_usafacts.pdf](https://github.com/cmu-delphi/covidcast-indicators/files/5549081/compare_with_usafacts.pdf)
- The geo-wise correlation is shown like this
![image](https://user-images.githubusercontent.com/31444565/99296378-a8b7b200-2814-11eb-93e2-d2dd1b9f76c1.png)
- Fixed state-level [visual view ](https://drive.google.com/file/d/1tuIdRg4JxbXYVoDHDIhbUnvOtUKJL0zK/view?usp=sharing)


So according to the comparison with incidence numbers in USA-Facts, we should treat their(nchs's) weekly report as incidence numbers. There are NAs existing in the raw dataset. I treat them as 0 when initially write the pipeline. But it is incorrect especially for states with a lot of missing values such as "AK", "DE", "HI", "ME", "NC", "WY" for either a certain time range or the entire published time range. I should keep those NAs, definitely. 
If keeping all of the NAs, and drop regions where one of the signal(nchs_mortality deaths_covid_num, usa-facts deaths_incidence_num) has an NA when calculating the spearman correlation, the result would show like this
![image](https://user-images.githubusercontent.com/31444565/99300494-abb5a100-281a-11eb-8150-1c58d3cf3eb7.png)
In W11, only WA has 27 reported deaths in NCHS mortality, only WA has 24 deaths and LA has 1 deaths in USA-Facts.

It's correct that they have 74 cases in Table 2 for AK, but it is also correct that they only have 0 or NA for AK in table 1 all the time (similar thing happens to WY, the first non-zero/non-NA value shows in W42 with week-ending date 2020-10-17). @RoniRos You can directly download the most recent raw dataset from [here](https://data.cdc.gov/NCHS/Provisional-COVID-19-Death-Counts-by-Week-Ending-D/r8kw-7aab) (click "export" and then "CSV"). So, it seems they have 74 recorded COVID related deaths until 11-14 but not sure which week they should assign those deaths to.

**Comment by RoniRos** [comment]
I think I figured out the reason for the discrepancy: any time the value is missing (N/A), it corresponds to a positive count less than 10 (i.e., somewhere in 1-9), which is censored by privacy rules.  For low-population, low-covid states like Alaska or Wyoming, this happens most of the time.  For some of the other states, it happens occasionally.
We need to figure out how to handle this.  These counts are not completely missing -- they may be reconstructed, at least partially, from NCHS cumulative values (Table 2, which gave us 74 cases for AK).  Have we been storing prior versions of Table 2?  @brookslogan @jingjtang ?

**Comment by jingjtang** [comment]
Didn't find a download link for Table2, but [here](https://data.cdc.gov/NCHS/Provisional-COVID-19-Death-Counts-in-the-United-St/kn79-hsxy) is a county level table for COVID-related deaths only. According to their [API docs](https://dev.socrata.com/foundry/data.cdc.gov/kn79-hsxy), we should be able to specify the `data_as_of`, however, I cannot fetch the data in old versions. (I changed the `data_as_of` in `https://data.cdc.gov/resource/kn79-hsxy.json?data_as_of=2020-11-12T00:00:00.000`, but got nothing) @brookslogan Have you played with the Socrata API to get old versions of data?

**Comment by jingjtang** [comment]
[Draft](https://docs.google.com/document/d/1RLy4O-gtACVjLEVD_vxyPqvp9nWVhqEjoWAKi68RKpg/edit#) for the signal description pop-up text added under `Name: Percent of Expected Deaths`.

**Comment by krivard** [comment]
when I looked at Socrata before, `data_as_of` did nothing. [Here's a Slack thread](https://delphi-org.slack.com/archives/C013AH5N01E/p1599576261054800?thread_ts=1599331248.052300&cid=C013AH5N01E) where Logan explains that it's not a first-class feature of their API.

I'd forgotten about the frontend support for weeklies 😱  we may end up pushing this to 1.12 in the map, but at least we can release in the API with the current batch.

**Comment by benjaminysmith** [comment]
Summarizing the last few comments: 
- We are not currently able to reconstruct the data because of the API limitations. It sounds like the default way for us to handle these for now is to fill these with N/A. @jingjtang is that an accurate assessment? If that is the simplest and we still think the data is useful, we could start with this and follow up to see if we can reconstruct later.
- We will need to follow up with map support for weekly data. @tildechris as FYI.

**Comment by capnrefsmmat** [comment]
Will this signal have `time_type = "week"` in the API? If so, we'll need to update the API clients to support it as well, and I'll need to file an issue for that.

**Comment by jingjtang** [comment]
@benjaminysmith Yes, since currently we don't have another data source that can handle the missing values due to privacy issue, we just want to keep them as NAN, at least for now I think @krivard .
And a PR [here](https://github.com/cmu-delphi/covidcast-indicators/pull/535#pullrequestreview-533590390) for this.

**Comment by tildechris** [comment]
Regarding map support, we'd need to know an ETA of when this would be available in staging to schedule the work to pull it in.

**Comment by benjaminysmith** [comment]
API support sounds like a blocker for release. Filed a request for this in https://github.com/cmu-delphi/covidcast/issues/305

**Comment by krivard** [comment]
*`covidcast` client support.

The API server has no problem with a weekly time type, and neither do the low-level clients in `delphi-epidata`. This may be a topic for team leads this afternoon.

**Comment by RoniRos** [comment]
Representing the censored values as N/A seems like the most reasonable thing to do for now.  Eventually, we may want to distinguish them from truly missing.  Do our tools support different types of missingness?

**Comment by RoniRos** [comment]
Also, we need to think of how to represent on the map zero vs. censored vs. unknown.

**Comment by krivard** [comment]
Verdict from leads: Go ahead with the release in the API; the client and map will catch up.

@RoniRos We do not yet encode different kinds of missingness, it's on the list for December.

**Comment by korlaxxalrok** [comment]
@krivard This is on prod now. It should have delivered some data yesterday, but otherwise will run daily, build up its cache, and then output on Mondays.

**Comment by krivard** [comment]
The API docs got lost along the way but I recovered them in https://github.com/cmu-delphi/delphi-epidata/pull/315.

@jingjtang would you draft a mailing list announcement and drop it in a comment here? Then @Akvannortwick can polish it up for distribution and maybe we can announce this puppy tomorrow.

**Comment by Akvannortwick** [comment]
You can also draft an email [here](https://docs.google.com/document/d/1vM8WWxIlxcuBSNTCIl95pKEd6szHCXn_utMUE409sng/edit?usp=sharing) and I will make changes and suggestions.

**Comment by jingjtang** [comment]
@krivard @Akvannortwick Except for the descriptions for all related signals, if there anything else to be added? 

Besides, I made some comments under [cmu-delphi/delphi-epidata#315](https://github.com/cmu-delphi/delphi-epidata/pull/315)

**Comment by krivard** [comment]
Released today!
